src/sentry/ingest/transaction_clusterer/datasource/redis.py

```diff
@@ -121,3 +116,31 @@ def _should_store_transaction_name(event_data: Mapping[str, Any]) -> bool:
         return False
 
     return True
+
+
+def _bump_rule_lifetime(project: Project, event_data: Mapping[str, Any]) -> None:
+    from sentry.ingest.transaction_clusterer import rules as clusterer_rules
+
+    applied_rules = event_data.get("_meta", {}).get("transaction", {}).get("", {}).get("rem", {})
+    if not applied_rules:
+        return
+
+    # XXX(iker): we're accessing project options here, and it represents a high
+    # (read) load.  However, since we're cleaning the redis store often, it's
+    # the only way to fetch the rules.
+    # Not checking if new rules exist means other strings could be added to the
+    # rules storage, which can explode the size of a project config.
+    stored_rules = clusterer_rules._get_rules(project)
```

Comment:
Instead of fetching rules here (we cannot access postgres for every transaction), I would give the `RedisRuleStore` (not the other rule stores) a method to update its rules (see pseudocode below).

The concern that redis might not be have all the rules all the time is valid, but I think we can live with missing updates occasionally, for the sake of performance.

```python
def _bump_rule_lifetime():
    applied_rules: Set[str] = get_applied_rules(event_data)
    RedisRuleStore().update_last_seen(project, {r: now() for r in applied_rules})
    
class RedisRuleStore:
    # ...
    def update_last_seen(project, rules, ts):
        existing_rules = self.read()
        updates = {rule: last_seen for rule in rules.items() if rule in existing_rules}
        redis.hmset(redis_key, updates)
```

If necessary, we can optimize even further:

1. Don't even read from redis and just write last seens for any rule Relay gives us. But do not _merge_ these rules into project options.
2. Do the `hmset` call inside the existing lua script.

Reply:
> how do you know if something in Redis is actually a rule?

By not allowing redis rules to merge into project options if they're not already present in project options. I realize that's not how we designed the `CompositeRuleStore`, but that can be changed.

> we already are for the feature flag

That lookup is heavily cached AFAIK.

> I'd like to separate the business logic from the stores a bit

The stores _are_ business logic in the sense that they have a single purpose, so I would not care about that too much.

Reply:
> By not allowing redis rules to merge into project options if they're not already present in project options.

This is a good point, but I see this as an additional safety guard. Even if we add that check, we don't want to add plenty of stuff to Redis just because its throughput supports it -- we don't have unlimited storage. For that reason, I believe checking if the rule exists in Redis before writing into it as a required step. And, if we already check Redis, I don't see the need to do the change you suggest at this point (in the future we may get to more issues).